### PR TITLE
Request-ify synthesis of the main function for the @main attribute

### PR DESCRIPTION
**Pull request by slavapestov**
Now that EmittedMembersRequest is cached, we need to synthesize it when collecting all the members of the class. Use a request for this, instead of just doing it from the AttributeChecker.

Also add some SILGen tests to ensure that the main function is actually emitted; previously we didn't test for this.

This fixes the regression from https://github.com/apple/swift/pull/32581.

### CODE REVIEW

**Comment by slavapestov** [comment]
CC @compnerd

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Review by nate-chandler — APPROVED** [review]

File: test/SILGen/` [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// RUN: %target-swift-emit-silgen -parse-as-library %s
```
Comment (theblixguy) [inline test/SILGen/`:1]: This file is called (`` ` ``), is this expected?

Comment (slavapestov) [reply test/SILGen/`:1]: No, that's a mistake. oops

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

File: lib/Sema/TypeCheckAttr.cpp [hunk 2]
```diff
@@ -1843,7 +1843,15 @@ synthesizeMainBody(AbstractFunctionDecl *fn, void *arg) {
...
 }
 
-void AttributeChecker::visitMainTypeAttr(MainTypeAttr *attr) {
+FuncDecl *
+SynthesizeMainFunctionRequest::evaluate(Evaluator &evaluator,
+                                        Decl *D) const {
+  auto &context = D->getASTContext();
+
+  MainTypeAttr *attr = D->getAttrs().getAttribute<MainTypeAttr>();
+  if (attr == nullptr)
```
Comment (AnthonyLatsis) [inline lib/Sema/TypeCheckAttr.cpp:1852]: Should this be an assert instead?

**Pull request merged**